\section{Conclusion and Future Work}

In this work, we examined the problem of developing a FASTQ-aware \gzip index
creation utility and parallel reader that leverages those pre-built indexes to
speed up \gzip FASTQ I/O bottlenecks. 

Toward that goal, we developed index building (\ibuilder) and index reading
(\ireader) utilities. Both our \ibuilder and \ireader utilities expand upon a
proof-of-concept for developing a \gzip index called \zran. Our \ibuilder
creates sequence read-aware indexes over the \gzip FASTQ file for a
user-configurable number of sequences in each ``chunk.'' We showed in
\S\ref{sec:ibuilder} that by smartly choosing the size of these sequence chunks
for index creation, the resulting index files sizes can be fractions of 1\% of
the size of the original \gzip FASTQ file. Using the \ireader utility we
developed, we take these index files and divvy the work of reading the chunks of
sequences up between a user-configurable number of threads. We show that our
implementation outperforms a serial, non-sequence-aware baseline in reading the
FASTQ file into memory. 

Finally, we demonstrate our implementation's ability to answer genomics
questions, such as counting the number of occurrences of nucleotides within a
\gzip FASTQ file, in a parallel fashion. We show that, for only a slight
performance degradation from the parallel reader, we can indeed perform this
task in an order of magnitude faster runtime than a serial implementation using
\texttt{zcat} and other Unix tools.

Our code is located here:
\url{https://github.com/Josh-Tang112/CMSC701_Final_Project}
